test/unit/alphabet/composition/cartesian_composition_test.cpp

```diff
@@ -0,0 +1,681 @@
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+#include <gtest/gtest.h>
+
+#include <seqan3/alphabet/all.hpp>
+
+using namespace seqan3;
+
+/************** TUPLE INHERITANCE **********************/
+
+template <typename type1, typename type2>
+struct test_composition : public cartesian_composition<test_composition<type1, type2>, type1, type2>
+{
+    using base_t = cartesian_composition<test_composition<type1, type2>, type1, type2>;
+    using base_t::base_t;
+    using base_t::operator=;
+
+    using base_t::operator==;
+    using base_t::operator!=;
+    using base_t::operator<;
+    using base_t::operator>;
+    using base_t::operator<=;
+    using base_t::operator>=;
+};
+
+template <typename T>
+class cartesian_composition_test: public ::testing::Test
+{
+public:
+
+    T instance = T{value_1(T{}), value_2(T{})};
+    T zero_instance = T{decltype(value_1(T{})){}, decltype(value_2(T{})){}};
+    size_t tup_size{2};
+
+    // test_composition<dna4, dna5>
+    // -------------------------------------------------------------------------
+    auto value_1(test_composition<dna4, dna5> const &)
+    {
+        return dna4::G;
+    }
+    auto assignable_to_value_1(test_composition<dna4, dna5> const &)
+    {
+        return rna4::G;
+    }
+    auto value_2(test_composition<dna4, dna5> const &)
+    {
+        return dna5::G;
+    }
+    auto assignable_to_value_2(test_composition<dna4, dna5> const &)
+    {
+        return rna5::G;
+    }
+    auto values_to_cmp(test_composition<dna4, dna5> const &)
+    {
+        return std::make_tuple(/*low */dna4::A, dna5::A,
+                               /*mid */dna4::C, dna5::C,
+                               /*high*/dna4::T, dna5::T);
+    }
+
+    // qualified<dna4, phred42>
+    // -------------------------------------------------------------------------
+    auto value_1(qualified<dna4, phred42> const &)
+    {
+        return dna4::G;
+    }
```

Comment:
what is the point of these functions? Why not use the values directly?

Reply:
because depending on the composition that is tested the type of the value is different. How do I generically assign values otherwise (if I do NOT want the default or highest value in the alphabet)?